lib/mutant/matcher/method.rb

```diff
@@ -1,133 +1,154 @@
 module Mutant
   class Matcher
-    # Matcher for subjects that are a specific method
+    # Abstract base class for method matchers
     class Method < self
-      include Adamantium::Flat, Concord::Public.new(:env, :scope, :target_method)
-      include AST::NodePredicates
+      include AbstractType,
+              Adamantium::Flat,
+              Concord::Public.new(:scope, :target_method, :evaluator)
 
       # Methods within rbx kernel directory are precompiled and their source
       # cannot be accessed via reading source location. Same for methods created by eval.
       BLACKLIST = %r{\A(kernel/|\(eval\)\z)}.freeze
 
-      # Enumerate matches
+      SOURCE_LOCATION_WARNING =
+        '%s does not have valid source location, unable to emit subject'.freeze
+
+      CLOSURE_WARNING =
+         '%s is dynamically defined in a closure, unable to emit subject'.freeze
+
+      # Matched subjects
       #
-      # @return [Enumerable<Subject>]
-      #   if no block given
+      # @param [Env] env
       #
-      # @return [self]
-      #   otherwise
+      # @return [Enumerable<Subject>]
       #
       # @api private
-      def each
-        return to_enum unless block_given?
+      def call(env)
+        evaluator.call(scope, target_method, env)
+      end
+
+      # Abstract method match evaluator
+      #
+      # Present to avoid passing the env argument around in case the
+      # logic would be implemnented directly on the Matcher::Method
+      # instance
+      class Evaluator
+        include AbstractType,
+                Adamantium,
+                Concord.new(:scope, :target_method, :env),
+                Procto.call,
+                AST::NodePredicates
+
+        # Matched subjects
+        #
+        # @return [Enumerable<Subject>]
+        #
+        # @api private
+        def call
+          return EMPTY_ARRAY if skip?
 
-        if !skip? && subject
-          yield subject
+          [subject].compact
         end
 
-        self
-      end
+      private
 
-    private
+        # Test if method should be skipped
+        #
+        # @return [Truthy]
```

Comment:
I do not like that:

* Its not correct. Because it actually returns `Env` as `Env#warn` returns `Env`. If you found an implenentation of `Env#warn` that returns `Array` its a bug.
* There is no intention to return `Env` hence it should not be documented. Nobody should rely on that.